### PR TITLE
issue #2273 - implement variable length "valuePosition" in page storage

### PR DESCRIPTION
**Pull request by dk-github**
to allow for Entry keys of arbitrary length, while not incurring unnecessary overhead (as 99% of keys are typically quite short)

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by porunov — COMMENTED** [review]
@dk-github Thank you for the fix! Could you re-target your PR to `v0.5` branch?

File: janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java
Comment (li-boxuan) [inline janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java:116]: It took me some time to understand this method thoroughly. I think the usage of entryBuffer here is not straight-forward to understand, since you are kind of mixing the two use cases (reading and moving cursor v.s. purely reading) together in one method. Would be helpful if you can add some javadoc.

**Review by li-boxuan — COMMENTED** [review]
The code logic looks brilliant and good to me. I just left a few comments.

Can you also include your test in https://github.com/JanusGraph/janusgraph/issues/2273#issuecomment-749234583? Putting it in JanusGraphTest will be good I think.

File: janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPageUtils.java
Comment (dk-github) [reply janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPageUtils.java:96]: I could change it here, but not sure how it could help reuse it in BufferPage - here we have an "unpacked" entry and decide how we are going to pack, there we get the packed entry and need to find out how it WAS packed - so it needs to take the byte array etc...

File: janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPageUtils.java
Comment (dk-github) [reply janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPageUtils.java:115]: done

File: janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java
Comment (dk-github) [reply janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java:116]: done

**Comment by dk-github** [comment]
> Can you also include your test in [#2273 (comment)](https://github.com/JanusGraph/janusgraph/issues/2273#issuecomment-749234583)? Putting it in JanusGraphTest will be good I think.

I am not sure if we actually want to add it as a permanent test though - at least not in this form... What exactly do we want to test - saving values with SET cardinality? There probably are existing tests. That values in SET can be more than 127 bytes? This seems to have been a limitation unique to in-memory backend implementation... That it can be even bigger - how much bigger do we want to assert? 1Mb? Also it makes sense to assert across all backends more than just in-memory...

File: janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPageUtils.java
Comment (li-boxuan) [reply janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPageUtils.java:97]: What I noticed is, in BufferPage.java, you always retrieve `valPos` before retrieving `valPosSize`. Since you have `valPos` available, you could call `BufferPageUtils::computeValPosSize(int valPos)` to retrieve `valPosSize`.

The caller in this class should then call `computeValPosSize((array[start+i]).getValuePosition())` to retrieve `valPosSize`.

Comment (dk-github) [reply janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPageUtils.java]: ok, got you now - done

**Comment by dk-github** [comment]
@li-boxuan 

Resolved all previous comments, commented on why I came to think the repro code shouldn't become permanent test.

Please could you have another look, approve or comment further?

**Comment by dk-github** [comment]
one of the builds failed on unrelated test for a different backend- not related to this change
[ERROR] Tests run: 49, Failures: 1, Errors: 0, Skipped: 11, Time elapsed: 683.532 s <<< FAILURE! - in org.janusgraph.diskstorage.solr.CQLSolrTest

**Comment by farodin91** [comment]
> one of the builds failed on unrelated test for a different backend- not related to this change
> [ERROR] Tests run: 49, Failures: 1, Errors: 0, Skipped: 11, Time elapsed: 683.532 s <<< FAILURE! - in org.janusgraph.diskstorage.solr.CQLSolrTest

This a known issue.

File: janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java
Comment (li-boxuan) [inline janusgraph-inmemory/src/main/java/org/janusgraph/diskstorage/inmemory/BufferPage.java:121]: nitpick: don't put the curly brace on a new line, to be consistent with the convention used in the codebase.

**Review by li-boxuan — APPROVED** [review]
LGTM 👍 
I checked codecov and verified your changes are covered by your unit test. I am fine with either adding https://github.com/JanusGraph/janusgraph/issues/2273#issuecomment-749234583 or not adding.
I left a nitpick on code style. That being said, I am okay without further changes.

**Comment by dk-github** [comment]
thanks for the review @li-boxuan 
pls ping me if you'd like me to fix the curly - and I guess you would have to re-approve?

If there are no further reviews/comments, I will merge this next week based on "lazy consensus"

**Comment by farodin91** [comment]
> thanks for the review @li-boxuan
> pls ping me if you'd like me to fix the curly - and I guess you would have to re-approve?
> 
> If there are no further reviews/comments, I will merge this next week based on "lazy consensus"

I don’t think we need a re approve for such a change.

I don’t think you will merge with lazy consensus, because if one committer reviewed and one have worked on it, it should be okay. In any case, you think the code should be reviewed another time. Just ask and see if some is reviewing it.

**Comment by farodin91** [comment]
What could do is a rebase on the top master. We integrated two checks into our PR system.

**Comment by dk-github** [comment]
> # [Codecov](https://codecov.io/gh/JanusGraph/janusgraph/pull/2294?src=pr&el=h1) Report
> > Merging [#2294](https://codecov.io/gh/JanusGraph/janusgraph/pull/2294?src=pr&el=desc) ([d936872](https://github.com/JanusGraph/janusgraph/commit/d936872df4c815ef524a4c78013d90f45a886af8)) into [master](https://codecov.io/gh/JanusGraph/janusgraph/commit/d565c6353db9045644e51361ec758a09141efe5c?el=desc) ([d565c63](https://github.com/JanusGraph/janusgraph/commit/d565c6353db9045644e51361ec758a09141efe5c)) will **decrease** coverage by `22.59%`.

The findings of this robot don't make much sense to me... 

It wants me to give it permissions to my repo in order to show full report (sic!), so I cannot fully see where it went wrong with the files that I did actually change. But the fact that I changed 3 files, and it shows that 450 files were impacted tells me all I need to know about its usefulness :)

**Comment by dk-github** [comment]
Same goes for the other robot - codacy. We already had discussion about it before, where I expressed my opinion on the usefulness of the 99% of "issues" it raises :)

**Comment by dk-github** [comment]
> What could you do a rebase on the top master? We integrated two checks into our PR system.

Done. 

Not sure I understood your previous comment though - is there anything else I need to do before I can merge it?

**Comment by li-boxuan** [comment]
> > Merging #2294 (d936872) into master (d565c63) will decrease coverage by 22.59%.
> 
> The findings of this robot don't make much sense to me...

As you can see, the coverage now is +0.06%, which makes sense. The -22.59% was just an interim report because some tests were still running at that time. The final coverage report aggregated results of all GitHub actions.

I usually ask for a re-approval if I make further changes after approval, but I guess that is not enforced.

> Same goes for the other robot - codacy. We already had discussion about it before, where I expressed my opinion on the usefulness of the 99% of "issues" it raises :)

I wasn't involved in that discussion before, but I think the codacy comments are fair this time. If we find it too annoying, probably we can make it show a warning rather than a failure on GitHub? Not sure if that is possible.

**Comment by porunov** [comment]
> I wasn't involved in that discussion before, but I think the codacy comments are fair this time. If we find it too annoying, probably we can make it show a warning rather than a failure on GitHub? Not sure if that is possible.

We also can configure Codacy to skip certain comments which we find unuseful. Unfortunately there is no `warning` state yet. The feature request for that is here: https://github.com/isaacs/github/issues/1465

**Comment by farodin91** [comment]
> Not sure I understood your previous comment though - is there anything else I need to do before I can merge it?

https://github.com/JanusGraph/janusgraph/blob/master/docs/development.md#review-then-commit-rtc


For codecov, we could check for flag to require minimal 10 reports before updating the comment.
For codacy, i don't codacy reports all issues to PR. just the critical or error.

**Comment by farodin91** [comment]
We can change the behavior of codecov: https://docs.codecov.io/docs/codecovyml-reference#comment

**Comment by dk-github** [comment]
so I fixed the curly and pacified Codacy, all checks are green now

if no further comments, I will merge later today

**Pull request merged**
